summingbird-storm/src/main/scala/com/twitter/summingbird/storm/collector/AggregatorOutputCollector.scala

```diff
@@ -0,0 +1,131 @@
+import scala.collection.{ Map => CMap }
+import scala.collection.JavaConverters._
+import java.util.{ List => JList }
+
+/**
+ *
+ * AggregatorOutputCollector is a wrapper around the SpoutOutputCollector.
+ * AsyncSummer is used to aggregate the tuples.
+ * Different streams have separate aggregators and caches.
+ */
+class AggregatorOutputCollector[K, V: Semigroup](
+    in: SpoutOutputCollector,
+    summerBuilder: SummerBuilder,
+    summerShards: KeyValueShards,
+    flushExecTimeCounter: Incrementor,
+    executeTimeCounter: Incrementor) extends SpoutOutputCollector(in) {
+
+  // An individual summer is created for each stream of data. This map keeps track of the stream and its corresponding summer.
+  private val cacheByStreamId = MMap[String, AsyncSummer[(K, V), Map[K, V]]]()
+
+  // As the crushDown happens at a stream level. We have a mapping from stream to the messageIds.
+  // The messageIds are further aggregated to the level of summerShards.
+  // The Map keeps track of aggregated tuples' messageIds as a list.
+  private val streamMessageIdTracker = MMap[String, MMap[Int, MList[Object]]]()
+
+  /**
+   * This method is invoked from the nextTuple() of the spout.
+   * This is triggered with tick frequency of the spout.
+   */
+  def timerFlush(): Unit = {
+    val startTime = Time.now
+    cacheByStreamId.foreach {
+      case (stream, cache) =>
+        val tupsOut = cache.tick.map { convertToSummerInputFormat(_) }
+        emitData(tupsOut, stream)
+    }
+    flushExecTimeCounter.incrBy(Time.now.inMillis - startTime.inMillis)
+  }
+
+  private def convertToSummerInputFormat(flushedCache: CMap[_, _]): CMap[Int, CMap[_, _]] =
+    flushedCache.groupBy { case (k, _) => summerShards.summerIdFor(k) }
+
+  /*
+    The method is invoked to handle the flushed cache caused by
+    exceeding the memoryLimit, which is called within add method.
+   */
+  private def emitData(tuples: Future[TraversableOnce[(Int, CMap[_, _])]], streamId: String): JList[Integer] = {
+    val startTime = Time.now
+    val flushedTups = Await.result(tuples)
+    val messageIdsTracker = streamMessageIdTracker.getOrElse(streamId, MMap[Int, MList[Object]]())
+    val returns = flushedTups.map {
+      case (k, v) =>
+        val messageIds = messageIdsTracker.remove(k)
+        val list = new java.util.ArrayList[AnyRef](2)
+        list.add(k.asInstanceOf[AnyRef])
+        list.add(v.asInstanceOf[AnyRef])
+        callEmit(messageIds, list, streamId)
+    }
+    val result = new java.util.ArrayList[Integer]()
+    returns.filter(_ != null).foreach { result.addAll(_) }
+    executeTimeCounter.incrBy(Time.now.inMillis - startTime.inMillis)
+    result
+  }
+
+  /*
+   This is a wrapper method to call the emit with appropriate signature
+   based on the arguments.
+  */
+  private def callEmit(messageIds: Option[TraversableOnce[AnyRef]], list: JList[AnyRef], stream: String): JList[Integer] = {
+    (messageIds.isEmpty, stream.isEmpty) match {
```

Comment:
This should be like:
```scala
(messageIds, stream.isEmpty) match {
  case (None, true) => in.emit(list)
  case (None, false) => in.emit(stream, list)
  case (Some(ids), true) => in.emit(list, ids)
  case (Some(ids), false) => in.emit(stream, list, ids)
}
```
Of course, the ack/fail logic should expect the `TraversableOnce` instead of the `Option`.